virttest/virt_vm.py

```diff
@@ -805,14 +805,14 @@ def _get_address(self, index=0, ip_version="ipv4", session=None,
                                                      devs=devs,
                                                      session=session,
                                                      timeout=timeout):
-            self.address_cache.drop(mac_pattern % mac)
-
             nic_params = self.params.object_params(nic.nic_name)
             pci_assignable = nic_params.get("pci_assignable") != "no"
 
             if not (pci_assignable or nic.nettype == "macvtap"):
                 raise VMAddressVerificationError(mac, ip_addr)
 
+            self.address_cache.drop(mac_pattern % mac)
```

Comment:
If not consider the usage of macvtap or device passthrough, the logic would go like:
```python
if not utils_net.verify_ip_address_ownership(ip_addr, mac):
    raise VMAddressVerificationError(mac, ip_addr)
    self.address_cache.drop(mac_pattern % mac)
```
then it would never drop outdated cache, and there is nothing different than removing this line.

I can see the point, that is, during migration, there is guest downtime which disturb guests to pass the verification of ip's ownership. In this case, the cache would better not to be dropped because we would go the next attempt. Actually, the issue does not only happen during migration, but also for any paused guests.

IMO, the solution would be making cache to be dropped when the last attempt failed, or even never dropping them (that would cause overhead).
```

Reply:
Yes, I suggested not dropping it at all unless we use `pci_assignamble` or `macvtap`, because then we raise an exception and either the caller will treat it, or attempt to do so again (which can either result in the same situation, or the verification could pass and we are fine). We don't have a way to detect the last call, but maybe we can add a `env.post_process` method to drop invalid entries, which should do. I don't see a way to detect the last attempt as the user might catch the `VMAddressVerificationError` and attempt to re-run it later (eg. after the migration or pause is over).

Reply:
Hmm actually we could improve this to not to raise the exception and simply keep trying until the timeout, then we'd not need to run the `login` multiple times in `vm.reboot` and we'd be able to detect the last attempt. So yes, that would also work for me and perhaps would be even easier to use (no need to re-execute `login`, no need to adjust affected places) so let me know which way you prefer.

Reply:
Yes, it looks better, please have a try.

Reply:
Actually this is not (realistically) possible, because currently we:

```
wait_for_login
  -> wait_for_get_address(..., timeout=timeout, ...)    # this would work perfectly
  -> while ...:
         return self.login(..., internal_timeout, ...)    # this wouldn't work as instead of overall timeout internal_timeout is used
```

So basically the first `wait_for_get_address` is reliable and would allow to wait for the last attempt, the later `self.login` is executed in a loop (to avoid stalls on partially initialized sshd) so waiting for the last iteration won't work here as last iteration is not really a last iteration and we might be still sleeping/migrating at that point.

So what do you say, is it fine to use this approach (potentially keeping unverified ip addresses in cache) is acceptable? It works well (no Avocado-vt failure in >24h (only real bugs uncovered by this)) and all related cache entries are freed during `vm.destroy()` (qemu_vm around line 3106).

Reply:
Then I'd prefer not dropping the cache as well, even for the case of using macvtap or device passthrough. It should be safe because we always verify the addresses before using, and the only side effect is some of the failed reason would be changed from `VMIPAddressMissingError` to `VMAddressVerificationError` and I think it is acceptable.

Btw, there is something we can do: mark the cache block as staled (but not drop it, that means we can still use it), and run garbage collection at the end of the test, and that would prevent the staled cache block to be used by the next test. With that we also can make the management asyncable. However, I don't request it to be implemented by this PR, or even I'm not sure if we really have to that as it is more complicated, but it does can reduce some overhead.

Reply:
Yes, I can imagine a better solution, but keeping the entries throughout test execution should be safe enough for now. I'll skip the invalidation in all cases then (I wanted to be conservative...)